web/packages/teleterm/src/mainProcess/agentDownloader/fileDownloader.ts

```diff
@@ -0,0 +1,126 @@
+    );
+
+    let handler: ReturnType<typeof this.createDownloadHandler>;
+
+    try {
+      return await new Promise<void>((resolve, reject) => {
+        handler = this.createDownloadHandler(url, downloadDirectory, {
+          resolve: resolve,
+          reject: reject,
+        });
+        this.window.webContents.session.on('will-download', handler);
+        this.window.webContents.downloadURL(url);
+      });
+    } finally {
+      this.window.webContents.session.off('will-download', handler);
+    }
+  }
+
+  private createDownloadHandler(
+    url: string,
+    downloadDirectory: string,
+    {
+      resolve,
+      reject,
+    }: {
+      resolve(): void;
+      reject(error: Error): void;
+    }
+  ) {
+    const onDownloadDone = () => {
+      resolve();
+      this.window.setProgressBar(-1);
+      this.logger.info('Download finished');
+    };
+
+    const onDownloadError = (error: Error) => {
+      reject(error);
+      this.window.setProgressBar(-1, { mode: 'error' });
+      this.logger.error(error);
+    };
+
+    return (_: Event, item: DownloadItem) => {
+      const isExpectedUrl = item.getURL() === url;
+      if (!isExpectedUrl) {
+        // handle only expected URL
+        return;
+      }
+
+      item.on('updated', (_, state) => {
+        switch (state) {
+          case 'interrupted':
+            onDownloadError(new Error('Download failed: interrupted'));
+            break;
+          case 'progressing':
+            this.onProgress(item.getReceivedBytes(), item.getTotalBytes());
+        }
+      });
+
+      item.once('done', (_, state) => {
+        switch (state) {
+          case 'completed':
+            onDownloadDone();
+            break;
+          case 'interrupted':
+            // TODO(gzdunek): electron doesn't expose much information about why the download failed.
+            // Fortunately, there is a PR in works that will add more info https://github.com/electron/electron/pull/38859.
+            // Use DownloadItem.getLastReason() when it gets merged.
+            onDownloadError(
+              new Error(
+                `Download failed. Requested file may not exist or is temporarily unavailable.`
+              )
+            );
+            break;
+          case 'cancelled':
+            onDownloadError(new Error('Download was cancelled.'));
+        }
+      });
+
+      // Set the save path, making Electron not to prompt a save dialog.
+      // We don't have to check if the filename contains forbidden characters, they are escaped by Chromium.
+      // For example, downloading from the URL localhost:1234/%2Ftest ("/" encoded as %2F) gives _test filename.
+      const filePath = path.join(downloadDirectory, item.getFilename());
+      item.setSavePath(filePath);
+    };
+  }
+
+  private onProgress(received: number, total: number) {
+    const progress = received / total;
+    this.window.setProgressBar(progress);
```

Comment:
Does this work on your Linux VM? I tried it on Ubuntu 20 and the progress bar doesn't show up in both the dev and the prod version. Also, the app icon is not there for some reason.

Reply:
Eh, I didn't check that since I assumed that such simple API must work on all platforms.
Now I read the docs more carefully, and it turns out that progress bar is supported only in Unity desktop 🥲 

I think we will have to implement the progress bar in the UI anyway, but probably at some later stage? (hopefully before the release) I can add a task for it. I wouldn't want to add more stuff to this PR.

> Also, the app icon is not there for some reason.

Do you mean the prod version? I just checked 13.2.1 on Linux VM and the icon was there.

Reply:
Honestly it'd be best to incorporate the progress bar within the icons next to the steps, so that a ring fills up as the download progresses. But yeah, I don't think it's necessary here.

> Do you mean the prod version? I just checked 13.2.1 on Linux VM and the icon was there.

I tried it on a locally packaged app. idk, maybe something's wrong with my setup.